# AI Deep Learngn Coding


Baraye code neevshtane tamame chzi haei k yad gereftim shoma mitavanid az ketabkhane haye moteadedi estefade konid, SKlearn, tensorflow, Pytorch


Sklearn --> Traditional Machine learning sakhte shdoe (linear regresion, KNN, DT , Random forest, SVM ) va hamchnin feature engineeering (mohandesi vizhegi) , pipeline, cross valdiation va ....

hamchnin support mikrd az do class **MLPclassifier** va **MLPRegresssor**

taamme in chizae k ayd grfte bodim ,dakhele model config koni

```python
from skelarn.neural_networks import MLPRegressor

model = MLPRegressor(hidden_Layer_sizes=(10,10,3) , optimizer='adam' , activation = 'relu')

model.fit(x_train,y_train) #ya inke

from sklearn.model_selection import GridSearchCV

gs = GridSearchCV(model ,hyperparametrs,.....)

gs.fit()

pipeline ...
```

moshjkelesh od k in ketabkhone baraye neural networks va deep leanrign sakhte nashdoe balke dota ketabkhone ma mohemtr darim --> Tensorflow va pytorch


### Moghayese tensorflow va pytorch

Tensorflow ---> Google (deepmind) --> khafane, koli utilities , tensofrboard, tensorserving (server) production kh behtare


Pytorch --> Meta (facebook, instagram) --> kh khafane --> academic bishtar dooseshd aran, ejaze mie, memari haye jadid ro emtehan konid, dasteton baz mizare, sakht tar ham hast


Bayad joftesho bald bashi, vali yek kodomo entekhab mikoni expert (motekhasessesh beshi)





----------
# Tensorflow Library tutorial

```bsh
                 TensorFlow

                      |
 ------------------------------------------------
 |              |              |                |
Core          Keras        TensorBoard       Deployment


```


Tensorflow Core hamon tensor ha , calculation tensor ha , gpu acceleration va automatic differentiation fucntions.

Kera high level API hast, tamame in atavbe ro va amaliat haro kenare hamgzoahste -> kolesho yek calss k shoma rahat estefade konid

Tensorboard --> tool (abzar) grapg, training model, loss , ....

Tensorflow data (tf.data) > dataset biarid 

tensorflow lite --> baraye edge device (mobile ,...)

tensorflow serving --> production  , API user bian daatresi 


```bsh
pip install tensorflow
```


```python

import tensorflow as tf
print(tf.__version__)

```


Tensor -> numpy array --> Yani yek chizi has . 0 bodi (adad) , 1 body (vector) , 2 body (matrix) , 3bodym,.-->tensoir


Input layer -> X = p10,20,30 , yek tensor namayehs bedim

weight hidden layer -> Tensor

Output --> Tensor


neural netwrok --> Flow i darim az input tensor , hidden layer tensor --> output tensor

Flowi az tensor --> TensorFlow


kole dade hat , oon array, oon operation ...  <br>

CPU --> 5 core , 7 core --> tedade core kam dre, parallel kardn kame , core kh ghavie 5 ta kare kh bzoorg ro ham zaman anjam bdi  --> default 

GPu --> kh core haye kochiki dare 1000 -->  kar haye kochik --> input * weight --> 1 million weight , x*weight --> operationn

1000 taro sanie baham parallel . ---> GPU --> shetab mdie training, estefade az DL

```python
import tensorflow as tf

tf.config.list_physical_devcies('gpu')

```
kjhorji b shoma 0 --> nist , 1 --> gpu_1 --> physical devcie_gpu1 





```python
import numpy as np

x = np.array([1,2,3])
```

```python
import tensorflow as tf

x = tf.constant([1,2,3])
```

Rank

#### Rank 0 - scalar

```python
x = tf.constant(5)
```


#### Rank 1 - Vector 
Yek bordar hat 
```python
x = tf.constant([1,2,3])
```

#### Rank 2 - Matrix 
So here is in 2 diemsion liek table (jadval)

```python
x = tf.constant(
[
[1,2],
[3,4]
]
)
```

Karborde aslish dar weight (vazn ) haye yek neurla network


#### Rank 3 - Tensor
like RGB which is height x width x channel  




Do ta ravesh yta constant besazid ya variable. constyant sabete --> taghir pazir nsitan , ama weight ha ma mikahhm weigth ha betonan taghir konan --> varibale


```python

x = tf.constant([1,2,3])
```

```python
x = tf.constant(
    [1, 2, 3],
    dtype=tf.float32
)
```
#weigth ha aksaran float hastan, 32 bit 


badan cast anjam bdi , taghir 

```python
x=tf.constant([1,2,3])

x=tf.cast(x, tf.float32)

```











```python
import tensorflow as tf

x= tf.varibale(5)

```

baraye taghir

```python
x.assign(10)
```




### Mathematcial oeprations on tensors



In [ ]:
import tensorflow as tf 

print(tf.__version__)

#python -m pip install --upgrade protobuf

2.21.0


In [5]:
import tensorflow as tf 

a = tf.constant([1,2,3])
b = tf.constant([4,5,6])

In [9]:
c  = a + b 

print(c)

tf.Tensor([5 7 9], shape=(3,), dtype=int32)


In [ ]:
c = a*b #element wise product

print(c)

tf.Tensor([ 4 10 18], shape=(3,), dtype=int32)


In [14]:
a = tf.constant([[1,2],[4,5]])
b = tf.constant([[4,5],[7,8]])

#(m*n)  (n*p) --> (m*p)


c = tf.matmul(a,b) #matrix multiplication    W*x   bayad weigth ha zarb va bad summ

print(c)

tf.Tensor(
[[18 21]
 [51 60]], shape=(2, 2), dtype=int32)


In [16]:
a = tf.constant([[1,2],[4,5]])

x = tf.reshape(a, (1,4))

print(x)

tf.Tensor([[1 2 4 5]], shape=(1, 4), dtype=int32)


In [18]:
x_epand = tf.expand_dims(x, axis=0)
print(x_epand)

tf.Tensor([[[1 2 4 5]]], shape=(1, 1, 4), dtype=int32)


In [ ]:
#broadcasting
a = tf.constant([1,2,3])

b = a + 5 

#bia b har elemntam 5 to ezafe kon

print(b)

#bias + weihht*Xa 


tf.Tensor([6 7 8], shape=(3,), dtype=int32)


In [21]:
x = tf.constant([1,2,3])

x_nump = x.numpy()

print(x_nump)

print(type(x_nump))









[1 2 3]
<class 'numpy.ndarray'>


In [23]:
#mesle numpy yeksri tabe haye geenrator ham dare

print(tf.zeros([2,3]))

print(tf.ones([2,3]))


x=tf.fill([2,3],5)

print(x)


print(tf.random.normal([2,3], mean=0, stddev=1))




tf.Tensor(
[[0. 0. 0.]
 [0. 0. 0.]], shape=(2, 3), dtype=float32)
tf.Tensor(
[[1. 1. 1.]
 [1. 1. 1.]], shape=(2, 3), dtype=float32)
tf.Tensor(
[[5 5 5]
 [5 5 5]], shape=(2, 3), dtype=int32)
tf.Tensor(
[[ 0.6644719   1.59703     0.38520354]
 [-2.186645    1.0078833   0.0950627 ]], shape=(2, 3), dtype=float32)


In [24]:
print(x.shape)
print(x.dtype)
print(tf.rank(x)) #dimension
print(tf.size(x))








(2, 3)
<dtype: 'int32'>
tf.Tensor(2, shape=(), dtype=int32)
tf.Tensor(6, shape=(), dtype=int32)


In [26]:
#indexing
x=tf.constant([10,20,30,40,50])

x[0] #getting one element


<tf.Tensor: shape=(), dtype=int32, numpy=10>

In [27]:
x[1:4] #getting slice

<tf.Tensor: shape=(3,), dtype=int32, numpy=array([20, 30, 40], dtype=int32)>

The matrix it has and also we can have them for instance

```bsh
1 2 3
4 5 6
```

```bsh
1 4
2 5
3 6
```


```python
tf.transpose(x)
```


Indexing , slice

```python
x=tf.constant([10,20,30,40,50])

x[0] #getting one element

x[1:4] #getting slice


```


Also concatenate , 
```python
tf.concat([a,b],axis=0)
```



also some operations we have


also we have some **Reducton Functions** :
these are important for loss and metrics. so for isnatnce 

```python
x=[1,2,3]

tf.reduce_sum(x) #6
```


it give us mean 
```python
tf.reduce_mean([1,2,3]) #2
```
for isnatnce it used in Mean squared error

it giev you the maximum values
```python
tf.reduce_max(x)
```
and also lowest valeus
```python
tf.reduce_min(x)
```




Also we have soem comaprisonal fucntions

```python
tf.greater(
[1,5],
[2,3]
)
#[False,True]
```

```python
tf.equal(a,b)
```

for egenration of ranoms we have different

```python
tf.random.normal([3,3])
```



```python
tf.random.uniform([3,3])
```

we can also set the random seed
```python
tf.random.set_seed(42)
```



In [30]:
import tensorflow as tf   

x= tf.constant([10.,20.,30.,-10.,-20.,-30.,0.])
z_x  = tf.nn.relu(x)

print(z_x)

#relu --> if x is positive return x , if x is negative (==0) return 0




tf.Tensor([10. 20. 30.  0.  0.  0.  0.], shape=(7,), dtype=float32)


In [32]:
z_x = tf.nn.sigmoid(x)

print(z_x)

tf.Tensor(
[9.9995458e-01 1.0000000e+00 1.0000000e+00 4.5397861e-05 2.0611537e-09
 9.3576229e-14 5.0000000e-01], shape=(7,), dtype=float32)


In [ ]:
x = tf.constant([100.0 , 40.0 , 60.0])

z_x = tf.nn.softmax(x)

print(z_x)

tf.Tensor([1.000000e+00 8.756511e-27 4.248354e-18], shape=(3,), dtype=float32)


Keras estefade mikonid khdoesh omde tensorflow haro tamamn compact krde too tavabe ei , shoma keraso yad bgiri lazem nis biay ba tensor ha kar koni



--------
---------
---------

# Pytorch  basic tutorials 


Meta (facebook, )


```bsh
pip install torch
```


In [35]:
import torch 

print(torch.__version__)

2.3.0


In [37]:
print(torch.cuda.is_available()) #Nvidia GPU

False


In [38]:
#tENSOR DARIM 
#0 bodu
import torch

x = torch.tensor(5)

print(x)

tensor(5)


In [39]:
x = torch.tensor([1, 2, 3, 4])

print(x)

tensor([1, 2, 3, 4])


In [40]:
print(x.ndim)

1


In [41]:
print(x.shape)

torch.Size([4])


In [42]:
x = torch.tensor([
    [1, 2, 3],
    [4, 5, 6]
])

print(x)

tensor([[1, 2, 3],
        [4, 5, 6]])


In [43]:
#3 body --> rgb 3 channel , pixel x pixel Red, Green, Blue
# 
x = torch.tensor([
    [
        [1, 2],
        [3, 4]
    ],
    [
        [5, 6],
        [7, 8]
    ]
]) 

In [44]:
#tavabe geenrator darim
x = torch.zeros(3, 4)

print(x)



tensor([[0., 0., 0., 0.],
        [0., 0., 0., 0.],
        [0., 0., 0., 0.]])


In [45]:
x = torch.ones(3, 4)

print(x)

tensor([[1., 1., 1., 1.],
        [1., 1., 1., 1.],
        [1., 1., 1., 1.]])


In [46]:
#beyne 0 o 1

x = torch.rand(3, 4)

print(x)


tensor([[0.9723, 0.7113, 0.9765, 0.6150],
        [0.0536, 0.2289, 0.3236, 0.3893],
        [0.1535, 0.0222, 0.8497, 0.9655]])


In [47]:
#az normal estefade mikone, miu = 0 , sigma = 1

x = torch.randn(3, 4)
print(x)

tensor([[ 0.9132, -0.4441,  0.6631,  1.1420],
        [ 0.5997, -1.5329,  0.5690, -0.9699],
        [-0.7990, -1.0804, -1.9225,  1.6691]])


In [48]:
x = torch.arange(0, 10)

print(x)

tensor([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])


In [49]:
x = torch.linspace(0, 1, 5)

print(x)

tensor([0.0000, 0.2500, 0.5000, 0.7500, 1.0000])


In [50]:
#dtype oon adadeton ahm doros koni, weight = float32
x = torch.tensor(
    [1, 2, 3],
    dtype=torch.float32
)

In [51]:
#amaliate riazi 

a = torch.tensor([1, 2, 3])
b = torch.tensor([4, 5, 6])

a+b

tensor([5, 7, 9])

In [52]:
a*b

tensor([ 4, 10, 18])

In [53]:
torch.matmul(a, b)

tensor(32)

In [54]:
a @ b

tensor(32)

In [55]:
# W*X + B   

a = torch.tensor([
    [1., 2.],
    [3., 4.]
])

b = torch.tensor([
    [5., 6.],
    [7., 8.]
])

c = a @ b

print(c)

tensor([[19., 22.],
        [43., 50.]])


In [ ]:
#reshape ,.....
#squeez()
#unsqueez()
#traspose

#x.min()   x.max() x.mean() x.sum()

In [57]:
output = torch.tensor([
    0.1,
    0.3,
    0.2,
    0.5,
    0.4
])


output.argmax()


tensor(3)

In [58]:
device = torch.device(
    "cuda" if torch.cuda.is_available() else "cpu"
)

print(device)

cpu


In [59]:
x = torch.tensor([1,2,3], device=device)
print(x)




tensor([1, 2, 3])


In [61]:
x = torch.tensor([1,2,3])

x = x.to(device)

print(x)

#harchizi k b in tensor marbote tamamn rooye on device man malom mikonm anajm mise


tensor([1, 2, 3])


In [62]:
import torch

device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')

print(device)



#--------------------------------
#-----Input
#--------------------------------
#yek x hast k daram
x = torch.tensor([2.0,3.0,4.0] , dtype=torch.float32)

#--------------------------------
#-----weights----
#--------------------------------
#yek alyer 3 weight, 1 neuron 

#3 ta vorodi --> neuron -->

w = torch.tensor([0.1,0.2,0.3] , dtype=torch.float32)


#--------------------------------
#-----bias------
#--------------------------------
b = torch.tensor(0.5, dtype=torch.float32)


#----
#---Move to device
#-----

x = x.to(device)
w = w.to(device)
b = b.to(device)


#--------------------------------
#-----Forward pass
#--------------------------------
z = w@x + b 

print(z) #neuron 

'''


X1 -
    - 
     -
      - >
X2 -----> nEURON (+B) --->Output
      ->
    -
   -
x3 -


''' 

cpu
tensor(2.5000)


'\n\n\nX1 -\n    - \n     -\n      - >\nX2 -----> nEURON (+B) --->Output\n      ->\n    -\n   -\nx3 -\n\n\n'

Pytorch yek chizi dare kh kh mohem hast , va aslan in hast k ma baramon baes mishe ba tesnorflow ahat frgh kone --> autograd

y = x**2 

dy/dx = 2**x

x = 3 --> gradient --> 6 ta



w , L . ro L / ro W . 
agar weight ro taghir bdim loss cheghdr taghir mikone, gradient bar asase gradient manfi 

w new = w old - learning_rate * gradient loss / w


loss = y - y^  

Y^ = function ( sum (w*neuron + b))

neuron = function (sum ())

z = 2*y

y = 4* x 


moshtaghe z / x --> moshatghe z / y  * moshtaghe y / x 

dakhele z --> 2* y --> dyanmic graoh dahste bashe k chijori asan ta aln boode


In [64]:
import torch

x = torch.tensor(3.0 , requires_grad = True) #mohaseb ekon ch balaei saresh miad

y = x**2 

y.backward() #in miad az y ta harja k has hesab mikone moshtagh haro --> hesa
#chizi nmide

In [65]:
x.grad

tensor(6.)

In [66]:
import torch

x = torch.tensor(
    3.0,
    requires_grad=True
)

y = x ** 2

y.backward()

print("x =", x)
print("y =", y)
print("gradient =", x.grad)

x = tensor(3., requires_grad=True)
y = tensor(9., grad_fn=<PowBackward0>)
gradient = tensor(6.)


In [68]:
x = torch.tensor(3.0 , requires_grad = True) #mohaseb ekon ch balaei saresh miad

y = x **2 

z = y + 5 


'''
dynamic graph

x = 3 
 |
 | squared 
 |
y=9 
 |
 | add 5
 |
z = 14
'''

z.backward()

In [69]:
x.grad 

tensor(6.)

Baraye tamame parametr haee k required_grad dare in kar ro mikone doostan


vaghty shoma feed forwar d--> randomized weigths * x ha mishe --> da rnahayat output -_> loss migire

gradient loss b har weight, weight layer akhar w1 , w2  . layer 2 hidden layeer, 1 hidden layer, x ha b hidden layer. 

bayad bdone Y chijori skahte shdoe az w* x haye ghabli ta done done btone moshatghh ro bgire



--------------
---------------
# Breats Cancer (Saratane sine)


yek neural network amozesh bedim k adad ha ya vazn hash jori bashan va learn shdoe bashan (train) k vaghty vorodi yek resulte breast cancer --> badkhime ya khoskhijme

In [70]:
from sklearn.datasets import load_breast_cancer
data = load_breast_cancer()

X = data.data
y = data.target

In [71]:
print(X.shape)
print(y.shape)






(569, 30)
(569,)


569 bimar dairm, k harkodom (yedone biamr. ) baray eoon saratane 30 ta feature ehsab shode to yek jadval , radius on toode, roshanaei ,...

feature1 feature2 featrue 3    ..... ...   output 


badkhim khsohklim 0 1 

In [81]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42,stratify=y)



from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)





In [ ]:
import tensorflow as tf  
from tensorflow import keras  
from tensorflow.keras import layers 


#30 (input) --> 64 (hidden layer 1) -> 32 (hidden layer 2) --> 1 (output) 


model = keras.Sequential([
      layers.Input(shape=(30,)) ,

      layers.Dense(64,activation='relu') , 

      layers.Dense(32,activation='relu'),

      #layers.Dense(2,activation ='softmax') --> 0.2  , 0.8 
      layers.Dense(1,activation = 'sigmoid') 

      #agar binary --> 1 neuron -> sigmoid 
      #agar regression --> 1 neuron --> Linear (relu)
      #agar multi class --> class neuron --> softmax 

])

#fahme activation ha dar ech komaki 


In [83]:
#roo datahye kam adam

#sgd

#sgd + momentum --> Yke adadi local minima farar ...

#adamW

#loss -->MSE, MAE --> regression, bianery BCE , multi class
model.compile(optimizer = 'adam' ,loss='binary_crossentropy'  ,metrics=['accuracy'])



In [84]:
#ta alamn config krdm

history = model.fit(X_train,y_train , epochs=50 , batch_size = 32 ,validation_split=0.2)


#546 data darim dostan  model = .. -> random weight bias
#546 * 0.8 = 436.8 --> 436  training data darim

#32 mifrese too feed forwar d, loss --> update weight haro , . inaghd anjam ide --> kole dataharo bbine 

#436/32 = 13.6 -> 14 iteration mizane ta kole dataro bbine --> 1 poch


#50 bar inkaro kon

#vliadation --> training ham tabdil valdiatuio




Epoch 1/50
12/12 ━━━━━━━━━━━━━━━━━━━━ 1s 11ms/step - accuracy: 0.5879 - loss: 0.6168 - val_accuracy: 0.8352 - val_loss: 0.4721
Epoch 2/50
12/12 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.9286 - loss: 0.4137 - val_accuracy: 0.9670 - val_loss: 0.3211
Epoch 3/50
12/12 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9615 - loss: 0.2775 - val_accuracy: 0.9780 - val_loss: 0.2188
Epoch 4/50
12/12 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.9643 - loss: 0.1886 - val_accuracy: 0.9780 - val_loss: 0.1571
Epoch 5/50
12/12 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9643 - loss: 0.1375 - val_accuracy: 0.9780 - val_loss: 0.1208
Epoch 6/50
12/12 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.9725 - loss: 0.1086 - val_accuracy: 0.9780 - val_loss: 0.0984
Epoch 7/50
12/12 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.9753 - loss: 0.0902 - val_accuracy: 0.9780 - val_loss: 0.0841
Epoch 8/50
12/12 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.9753 - loss: 0.0798 - val_accuracy: 0.9780 - val_loss

In [85]:
test_loss , test_accuracy = model.evaluate(X_test,y_test)

print('test loss : ',test_loss)
print('test accuracy : ',test_accuracy)

4/4 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9561 - loss: 0.1195 
test loss :  0.11945711076259613
test accuracy :  0.9561403393745422


In [86]:
predictions = model.predict(X_test)

print(predictions[:5])

4/4 ━━━━━━━━━━━━━━━━━━━━ 0s 9ms/step 
[[3.5209441e-13]
 [9.9999958e-01]
 [1.3639565e-05]
 [4.2342011e-02]
 [2.3540197e-16]]


In [ ]:
#datato --> too --> weight, bias --> 0 - 1 (sigmoid)  ehtemale badkhim boidna gast

## Pytorch


In [87]:
from sklearn.datasets import load_breast_cancer

data = load_breast_cancer()
X = data.data
y = data.target


from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42,stratify=y)


from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)





In [ ]:
import numpy as np
import torch


#numpy array --> tensor 

X_train = torch.tensor(X_train, dtype=torch.float32)
X_test = torch.tensor(X_test, dtype=torch.float32)
y_train = torch.tensor(y_train, dtype=torch.float32)
y_test = torch.tensor(y_test, dtype=torch.float32)


#pytorch dataloader

batch_size = 32

from torch.utils.data import TensorDataset
#x , y train --> train_dataset
train_dataset = TensorDataset(X_train, y_train)
test_dataset = TensorDataset(X_test, y_test)

#dataset

#dataset --> shuffl (random) --> batch size --> Dataloader

from torch.utils.data import DataLoader

train_loader = DataLoader(train_dataset, batch_size=batch_size, shuffle=True)
test_loader = DataLoader(test_dataset, batch_size=batch_size, shuffle=False)

#batch batch data bgrim, 32 32 ta data .

#for gharar hast bznm , for x , y in data loader 

In [ ]:
#from torch import nn 
import torch.nn as nn

#yek ravesh shabihe rtesnor flow
class NeuralNetwork(nn.Module):
    def __init__(self):
        super().__init__()


        self.network = nn.Sequential(

            nn.Linear(30,64),
            nn.ReLU(),
            nn.Linear(64,32),
            nn.ReLU(),
            nn.Linear(32,1),
        )
        

    def forward(self,x):
        return self.network(x)




class NeuralNetwork(nn.Module):
    def __init__(self):
        super().__init__()
        #laye hamo
        #input(30)  ,dense(64)

        # onnection 
        self.layer1 = nn.Linear(30,64)

        self.relu1 = nn.ReLU()

        self.layer2 = nn.Linear(64,32)
        self.relu2 = nn.ReLU()

        self.output = nn.Linear(32,1)

        #sigmoid bezarid --> binary --> Loss havaseton 

        #multi clas -> softmax

        #regression -> linear (activation = none)

    def forward(self,x):
        x = self.layer1(x)
        x = self.relu1(x)

        x = self.layer2(x)
        x = self.relu2(x)

        x = self.output(x)

        return x


    

In [92]:
model = NeuralNetwork()

In [93]:
criterion = nn.BCEWithLogitsLoss() #BEHET ADAD midam , sigmoid nadaram 0 1 


In [94]:
optimizer = torch.optim.Adam(model.parameters(), lr=0.001)


/Users/apm/anaconda3/envs/DL/lib/python3.10/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


In [96]:
epochs = 50 

for epoch in range(epochs):
    for x_bath, y_bath in train_loader:

        #feedforward beram
        # 
        output = model(x_bath) 

        #loss 
        loss = criterion(output,y_bath.reshape(-1,1))
        #fasleye psihbini vaghheiatyo dar biar


        #clear old gradient haye ghabli

        optimizer.zero_grad()

        loss.backward()  #moshtagh haro hesab kon . ro loss / ro Weight

        optimizer.step() # weight new = weight old - learning_rate * gradient loss / weight


    print(f'Epoch {epoch+1}/{epochs}, Loss: {loss.item():.4f}')


        

Epoch 1/50, Loss: 0.5917
Epoch 2/50, Loss: 0.4260
Epoch 3/50, Loss: 0.2422
Epoch 4/50, Loss: 0.1335
Epoch 5/50, Loss: 0.4799
Epoch 6/50, Loss: 0.0822
Epoch 7/50, Loss: 0.0812
Epoch 8/50, Loss: 0.0207
Epoch 9/50, Loss: 0.0152
Epoch 10/50, Loss: 0.1072
Epoch 11/50, Loss: 0.0444
Epoch 12/50, Loss: 0.0299
Epoch 13/50, Loss: 0.0033
Epoch 14/50, Loss: 0.0029
Epoch 15/50, Loss: 0.0118
Epoch 16/50, Loss: 0.0166
Epoch 17/50, Loss: 0.0019
Epoch 18/50, Loss: 0.0075
Epoch 19/50, Loss: 0.0127
Epoch 20/50, Loss: 0.0185
Epoch 21/50, Loss: 0.0387
Epoch 22/50, Loss: 0.0005
Epoch 23/50, Loss: 0.0076
Epoch 24/50, Loss: 0.0387
Epoch 25/50, Loss: 0.0086
Epoch 26/50, Loss: 0.4276
Epoch 27/50, Loss: 0.0008
Epoch 28/50, Loss: 0.0016
Epoch 29/50, Loss: 0.0162
Epoch 30/50, Loss: 0.0549
Epoch 31/50, Loss: 0.0813
Epoch 32/50, Loss: 0.0650
Epoch 33/50, Loss: 0.0095
Epoch 34/50, Loss: 0.0036
Epoch 35/50, Loss: 0.0013
Epoch 36/50, Loss: 0.0052
Epoch 37/50, Loss: 0.0481
Epoch 38/50, Loss: 0.0337
Epoch 39/50, Loss: 0.

In [97]:
model.eval() #evaluate 

with torch.no_grad(): #gradient hesab nakoni
    logits = model(X_test) #x_test --> 3d3232

    probabilities = torch.sigmoid(logits)

    predictions = (probabilities >= 0.5).float()

    acccuracy = (predictions == y_test.reshape(-1,1)).float().mean()

    #3433 --> sigmoid --> 0 - 1 --> miam bala tar az 0,5 1 ,  0 

    #y tst -> 0 1 

    #chnataro doros psihbni

print(f'Test Accuracy: {acccuracy:.4f}')





Test Accuracy: 0.9561
